src/internals/match/Match.ts

```diff
@@ -0,0 +1,44 @@
+import { Fn, unset, _ } from "../core/Core";
+import { Functions } from "../functions/Functions";
+import * as Impl from "./impl/match";
+
+/**
+ * Match creates a pattern matching expression.
+ * Pattern matching let you execute different
+ * branches of code based on the structure
+ * of the input type exdestructure the input type, and execute
+ *
+ * @param input - the input to pattern match on
+ * @param withClauses - list of with clauses representing different branches if code
+ *
+ * @example
+ * ```ts
+ * type Test<T> = Match<T, [
+ *   With<{ msg: string }, F.ComposeLeft<[O.Get<"msg">, S.Prepend<"Message: ">]>>,
+ *   With<string, S.Append<" <-- Message">>,
+ *   With<any, F.Constant<"default value">>
+ * ]>
+ * ```
+ */
+export type Match<
+  valueOrWithClauses = unset,
+  withClauses = unset
```

Comment:
maybe add constraints to have errors when suplying an array of With clause
```suggestion
  withClauses = Impl.With<unknown, any>[] | _ | unset
```